src/Configuration/Programmer/Programmer_cmd.cpp

```diff
@@ -19,57 +19,140 @@ You should have received a copy of the GNU General Public License
+  CFG_POST_MSG("-- GetAvailableCables(cables) --")
+  for (auto& cable : cables) {
+    CFG_POST_MSG("--------------")
+    CFG_POST_MSG("product ID: 0x%04x", cable.productId);
+    CFG_POST_MSG("vendor  ID: 0x%04x", cable.vendorId);
+    CFG_POST_MSG("serial    : %s", cable.serialNumber.c_str());
+    CFG_POST_MSG("desc      : %s", cable.description.c_str());
+    CFG_POST_MSG("bus       : %d", cable.busAddr);
+    CFG_POST_MSG("port      : %d", cable.portAddr);
+    CFG_POST_MSG("device    : %d", cable.deviceAddr);
+    CFG_POST_MSG("speed     : %d", cable.speed);
+    CFG_POST_MSG("transport : %d", cable.transport);
+    CFG_POST_MSG("channel   : %d", cable.channel);
+  }
+
+  CFG_POST_MSG("-- ListDevices(cables, outputMsg) --")
+  ret = ListDevices(cables[0], devices);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  }
+  if (devices.size() == 0) {
+    CFG_POST_MSG("No device found. Make sure the board power is ON.");
+    return -1;
+  }
+  CFG_POST_MSG("### ListDevices API testing ###");
+  for (auto& device : devices) {
+    CFG_POST_MSG("-------------------------");
+    CFG_POST_MSG("Device name: %s", device.name.c_str());
+    CFG_POST_MSG("Device index: %d", device.index);
+    CFG_POST_MSG("Device jtagId: 0x%08x", device.tapInfo.idCode);
+    CFG_POST_MSG("Device mask: 0x%08x", device.tapInfo.irMask);
+    CFG_POST_MSG("Device irlength: %d", device.tapInfo.irLen);
+    CFG_POST_MSG("Device flashSize: %s bits",
+                 CFG_convert_number_to_unit_string(device.flashSize).c_str());
     CFG_POST_MSG("-------------------------");
-    CFG_POST_MSG(outputMsg.c_str());
   }
 
-  Device device = devices[0];
   CfgStatus status;
-  if (GetFpgaStatus(device, status)) {
-    CFG_POST_MSG("### GetFpgaStatus API testing ###");
-    CFG_POST_MSG("Device cfgDone: %s", std::to_string(status.cfgDone).c_str());
-    CFG_POST_MSG("Device cfgError: %s",
-                 std::to_string(status.cfgError).c_str());
+  ret = GetFpgaStatus(cables[0], devices[0], status);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
   }
+  CFG_POST_MSG("### GetFpgaStatus API testing before programming ###");
+  CFG_POST_MSG("Device cfgDone: %d", status.cfgDone);
+  CFG_POST_MSG("Device cfgError: %d", status.cfgError);
 
-  OutputCallback callback = [](std::string msg) {
-    CFG_POST_MSG("### OutputCallback API testing ###");
-    CFG_POST_MSG("OutputCallback msg: %s", msg.c_str());
-  };
-  ProgressCallback progress = [](double percent) {
-    CFG_POST_MSG("### ProgressCallback API testing ###");
-    CFG_POST_MSG("ProgressCallback percent: %s",
-                 std::to_string(percent).c_str());
-  };
+  int userInput = 1;
   std::atomic<bool> stop = false;
-  ProgramFpga(device, "test.bit", "config.cfg", nullptr, callback, progress,
-              stop);
+  std::string bitfile = "/your/bitstream/file.bin";
```

Comment:
I don't understand this. Is that fake path and we don't care about real path? Why isn't it empty? 
If it matters it should be real path that could be read from some configuration file, for example. We can't push it like this to public repo.

Reply:
offline discussion to resolve this question